src/main/kotlin/org/digma/intellij/plugin/ui/tests/TestsServiceImpl.kt

```diff
@@ -0,0 +1,163 @@
+import org.digma.intellij.plugin.analytics.AnalyticsService
+import org.digma.intellij.plugin.analytics.AnalyticsServiceException
+import org.digma.intellij.plugin.analytics.EnvironmentChanged
+import org.digma.intellij.plugin.common.Backgroundable
+import org.digma.intellij.plugin.common.CommonUtils
+import org.digma.intellij.plugin.common.LOCAL_ENV
+import org.digma.intellij.plugin.common.LOCAL_TESTS_ENV
+import org.digma.intellij.plugin.common.isEnvironmentLocal
+import org.digma.intellij.plugin.common.isEnvironmentLocalTests
+import org.digma.intellij.plugin.common.isLocalEnvironmentMine
+import org.digma.intellij.plugin.document.CodeObjectsUtil
+import org.digma.intellij.plugin.errorreporting.ErrorReporter
+import org.digma.intellij.plugin.insights.InsightsModelReactHolder
+import org.digma.intellij.plugin.jcef.common.JCefMessagesUtils
+import org.digma.intellij.plugin.log.Log
+import org.digma.intellij.plugin.ui.jcef.serializeAndExecuteWindowPostMessageJavaScript
+import org.digma.intellij.plugin.ui.model.CodeLessSpanScope
+import org.digma.intellij.plugin.ui.model.EndpointScope
+import org.digma.intellij.plugin.ui.model.MethodScope
+import org.digma.intellij.plugin.ui.model.Scope
+import org.digma.intellij.plugin.ui.service.FillerOfLatestTests
+import org.digma.intellij.plugin.ui.service.FilterForLatestTests
+import org.digma.intellij.plugin.ui.service.ScopeRequest
+import org.digma.intellij.plugin.ui.service.TestsService
+import org.digma.intellij.plugin.ui.tests.model.EnvironmentEntity
+import org.digma.intellij.plugin.ui.tests.model.SetEnvironmentsMessage
+import org.digma.intellij.plugin.ui.tests.model.SetEnvironmentsMessagePayload
+
+class TestsServiceImpl(val project: Project) : TestsService {
+
+    private val logger = Logger.getInstance(this::class.java)
+
+    private var cefBrowser: CefBrowser? = null
+    private var fillerOfLatestTests: FillerOfLatestTests? = null
+
+    init {
+
+        project.messageBus.connect(this).subscribe(EnvironmentChanged.ENVIRONMENT_CHANGED_TOPIC, object : EnvironmentChanged {
+            override fun environmentChanged(newEnv: String?, refreshInsightsView: Boolean) {
+                // dp nothing
+            }
+
+            override fun environmentsListChanged(newEnvironments: MutableList<String>?) {
+                sendOperativeEnvironments()
+            }
+        })
+
+    }
+
+    override fun dispose() {
+        //nothing to do
+    }
+
+    override fun initWith(cefBrowser: CefBrowser, fillerOfLatestTests: FillerOfLatestTests) {
+        this.cefBrowser = cefBrowser
+        this.fillerOfLatestTests = fillerOfLatestTests
+    }
+
+    private fun scope(): Scope {
+        return project.service<InsightsModelReactHolder>().model.scope
+    }
+
+    override fun refresh() {
+        Backgroundable.ensurePooledThread {
+            val scopeRequest = getScopeRequest()
+            scopeRequest?.let {
+                fillerOfLatestTests!!.fillDataOfTests(cefBrowser!!, it)
```

Comment:
possible NPE. kotlin best practices, use  fillerOfLatestTests?.let with lambda arg name

Reply:
again , those 2 members are being set during the startup/init